ddprof-ffi/src/exporter.rs

```diff
@@ -275,8 +258,135 @@ pub unsafe extern "C" fn profile_exporter_send(
         Ok(HttpStatus(response.status().as_u16()))
     }() {
         Ok(code) => SendResult::HttpResponse(code),
-        Err(err) => SendResult::Failure(error_into_buffer(err)),
+        Err(err) => SendResult::Failure(err.into()),
+    }
+}
+
+#[export_name = "ddprof_ffi_SendResult_drop"]
+pub unsafe extern "C" fn send_result_drop(result: SendResult) {
+    std::mem::drop(result)
+}
+
+#[export_name = "ddprof_ffi_Vec_tag_new"]
+pub extern "C" fn vec_tag_new<'a>() -> crate::Vec<Tag<'a>> {
+    crate::Vec::default()
+}
+
+/// Pushes the tag into the vec.
+#[export_name = "ddprof_ffi_Vec_tag_push"]
+pub unsafe extern "C" fn vec_tag_push<'a>(vec: &mut crate::Vec<Tag<'a>>, tag: Tag<'a>) {
+    vec.push(tag)
+}
+
+#[allow(clippy::ptr_arg)]
+#[export_name = "ddprof_ffi_Vec_tag_as_slice"]
+pub extern "C" fn vec_tag_as_slice<'a>(vec: &'a crate::Vec<Tag<'a>>) -> Slice<'a, Tag<'a>> {
+    vec.as_slice()
+}
+
+#[export_name = "ddprof_ffi_Vec_tag_drop"]
+pub extern "C" fn vec_tag_drop(vec: crate::Vec<Tag>) {
+    std::mem::drop(vec)
+}
+
+struct ParseTagsError {
+    message: &'static str,
+}
+
+impl Debug for ParseTagsError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Tag Error: {}", self.message)
+    }
+}
+
+impl Display for ParseTagsError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "The tags string was malformed.")
+    }
+}
+
+impl Error for ParseTagsError {}
+
+fn parse_tag_chunk(chunk: &str) -> Result<Tag, ParseTagsError> {
+    if let Some(first_colon_position) = chunk.find(':') {
+        if first_colon_position == 0 {
+            return Err(ParseTagsError {
+                message: "Tag cannot start with a colon.",
+            });
+        }
+        let name = &chunk[..first_colon_position];
+        let value = &chunk[(first_colon_position + 1)..];
+        Ok(Tag {
+            name: CharSlice::from(name),
+            value: CharSlice::from(value),
+        })
+    } else {
+        Err(ParseTagsError {
+            message: "Unable to find a colon in the tag.",
+        })
+    }
+}
+
+/// Parse a string of tags typically provided by environment variables
+/// The tags are expected to be either space or comma separated:
+///     "key1:value1,key2:value2"
+///     "key1:value1 key2:value2"
```

Comment:
This is interesting; looking at the Ruby code, it really only allows commas. Can you point me in the direction where you found the space-separated version? I may need to bring this up with the rest of the Ruby folks if we're doing this wrong.

Reply:
Python previously just supported `,` and later added space separated to match the agent. Note that as far as I can tell, neither spaces nor commas are supported in tags: https://docs.datadoghq.com/getting_started/tagging/.

However, there is a flaw in this technique for choosing between them. Given tag `env:staging:east`, the key is `env` and the value is `staging:east` but neither method will split this correctly because it assumes only 1 colon is present per key-value chunk. Python has this same flaw.